telemetry/telemetry.go

```diff
@@ -0,0 +1,111 @@
+// As part of the White Glove White Glove project,
+// this is a telemetry package to support posting summary and cancellation events
+// to a given webhook endpoint with a reference project id in the shape of an event grid
+// custom event.
+
+package telemetry
+
+import (
+	"bytes"
+	"encoding/json"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/Azure/azure-storage-azcopy/common"
+)
+
+var (
+	// ReportingAPIURI the Webhook URI
+	ReportingAPIURI string
+	// ProjectReferenceID the project reference ID
+	ProjectReferenceID string
+)
+
+type telemetry struct {
+	JobID     string      `json:"jobID"`
+	ProjectID string      `json:"projectId"`
+	Telemetry interface{} `json:"telemetry"`
+}
+
+type reportEvent struct {
+	EventType  string      `json:"eventType"`
+	EventTopic string      `json:"eventTopic"`
+	Data       interface{} `json:"data"`
+	EventTime  time.Time   `json:"eventTime"`
+}
+
+func getJSON(t interface{}) string {
+	json, err := json.Marshal(t)
+	common.PanicIfErr(err)
+	return string(json)
+}
+
+func newReportEvent(eventType string, data interface{}) *reportEvent {
+	return &reportEvent{
+		EventType:  eventType,
+		EventTopic: "",
+		Data:       data,
+		EventTime:  time.Now(),
+	}
+}
+
+func postEvent(json string) {
+	var jsonStr = []byte(json)
+	_, err := http.Post(ReportingAPIURI, "application/json", bytes.NewBuffer(jsonStr))
+	common.PanicIfErr(err)
```

Comment:
Also, consider using the pipeline architecture in place at [https://github.com/Azure/azure-pipeline-go](https://github.com/Azure/azure-pipeline-go)

and preferably have some level of authentication on your telemetry service

Reply:
@adreed-msft What would be the advantages of using pipeline in this case (other than it being consistent with the rest of the app?).  Off the top of my head, there are not many things that the pipeline offers that are needed right here.  E.g. we wouldn't add logging to the pipeline here, since this already _is_ the logging.

Reply:
+1 re some kind of authentication, or, if it's assumed that that will be unnecessary because the telemetry URL will include a sas-like token, or be entirely within the organisation firewall where AzCopy is running?  Perhaps we should chat offline to discuss in more detail.